compiler/plugins/target/LLVMCPU/ResolveCPUAndCPUFeatures.cpp

```diff
@@ -0,0 +1,163 @@
+// Copyright 2020 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "compiler/plugins/target/LLVMCPU/ResolveCPUAndCPUFeatures.h"
+
+#include "llvm/TargetParser/Host.h"
+#include "llvm/TargetParser/RISCVTargetParser.h"
+#include "llvm/TargetParser/SubtargetFeature.h"
+#include "llvm/TargetParser/X86TargetParser.h"
+
+namespace mlir::iree_compiler::IREE::HAL {
+
+namespace {
+
+bool resolveHostCPUAndCPUFeatures(const llvm::Triple &triple, std::string &cpu,
+                                  std::string &cpuFeatures) {
+  if (cpu != "host" && cpuFeatures != "host") {
+    return true;
+  }
+  if ((!cpu.empty() && cpu != "host") ||
+      (!cpuFeatures.empty() && cpuFeatures != "host")) {
+    llvm::errs()
+        << "error: If either cpu or CpuFeatures is `host`, the other must "
+           "be either also `host` or the default value\n";
+    return false;
+  }
+  cpu = llvm::sys::getHostCPUName();
+  llvm::SubtargetFeatures features;
+  for (auto &feature : llvm::sys::getHostCPUFeatures()) {
+    features.AddFeature(feature.first(), feature.second);
+  }
+  cpuFeatures = features.getString();
+  return true;
+}
+
+bool logUnspecifiedCPUOrCPUFeatures(
+    const llvm::Triple &triple, std::string &cpu, std::string &cpuFeatures,
+    std::string_view loggingUnspecifiedTargetCPU) {
+  if (!cpu.empty() || !cpuFeatures.empty()) {
+    return true;
+  }
+  if (loggingUnspecifiedTargetCPU.empty()) {
+    return false;
+  }
+  if (triple.isX86()) {
+    // On X86, one usually talks in terms of CPU, not individual CPU features.
+    llvm::errs() << R"MSG(
+error: Please pass --iree-llvmcpu-target-cpu (or alternatively, --iree-llvmcpu-target-cpu-features, but this message is tailored to the current target architecture, x86, where --iree-llvmcpu-target-cpu is used more often).
```

Comment:
Considering the case where the user is compiling a module with these set via attributes and not via CLI flags, this should be an op diagnostic and not go directly to `stderr` via `llvm::errs()`.

Errors may not be as actionable when using a hosting framework that _should_ be setting the target CPU but isn't, so we may still want to just refer to these settings via their CLI flag names.

Reply:
All good points! With op diagnostics, can I generate a warning as opposed to an error?

I need to refocus a bit on GPU data tiling now, so we can give this some time to discuss and sort out exactly what we want to do. Maybe we can pick this up when you're back.  Converting back to draft for now.